File: cmd/igor/internal/feature/registry.go
package feature

// Registry holds the full catalog of available features.
type Registry struct {
	features map[string]*Feature
	order    []string // insertion order for deterministic iteration
}

// NewRegistry creates the built-in feature registry populated from
// the containers build-args schema and hand-maintained metadata.
func NewRegistry() *Registry {
	r := &Registry{
		features: make(map[string]*Feature),
	}

	// === Languages ===
	r.add(&Feature{
		ID: "python", BuildArg: "INCLUDE_PYTHON",
		DisplayName: "Python", Description: "Python runtime (auto-resolves to latest patch)",
		Category: CategoryLanguage, VersionArg: "PYTHON_VERSION", DefaultVersion: "3.14",
		EnvFile:      "python.env",
		CacheVolumes: []string{"pip-cache:/cache/pip", "poetry-cache:/cache/poetry"},
	})
	r.add(&Feature{
		ID: "python_dev", BuildArg: "INCLUDE_PYTHON_DEV",
		DisplayName: "Python Dev", Description: "Python development tools (linters, formatters, LSP)",
		Category: CategoryLanguage, IsDev: true, BaseLang: "python",
		EnvFile:          "python-dev.env",
		Requires:         []string{"python"},
		VSCodeExtensions: []string{"ms-python.python", "ms-python.vscode-pylance", "charliermarsh.ruff"},
	})

	r.add(&Feature{
		ID: "node", BuildArg: "INCLUDE_NODE",
		DisplayName: "Node.js", Description: "Node.js runtime (auto-resolves to latest patch)",
		Category: CategoryLanguage, VersionArg: "NODE_VERSION", DefaultVersion: "22",
		EnvFile:      "node.env",
		CacheVolumes: []string{"npm-cache:/cache/npm"},
	})
	r.add(&Feature{
		ID: "node_dev", BuildArg: "INCLUDE_NODE_DEV",
		DisplayName: "Node.js Dev", Description: "Node.js development tools (LSP, debug)",
		Category: CategoryLanguage, IsDev: true, BaseLang: "node",
		EnvFile:          "node.env",
		Requires:         []string{"node"},
		VSCodeExtensions: []string{"dbaeumer.vscode-eslint"},
	})

	r.add(&Feature{
		ID: "rust", BuildArg: "INCLUDE_RUST",
		DisplayName: "Rust", Description: "Rust toolchain (auto-resolves to latest patch)",
		Category: CategoryLanguage, VersionArg: "RUST_VERSION", DefaultVersion: "1.83",
		EnvFile:      "rust.env",
		CacheVolumes: []string{"cargo-cache:/cache/cargo", "rustup-cache:/cache/rustup"},
	})
	r.add(&Feature{
		ID: "rust_dev", BuildArg: "INCLUDE_RUST_DEV",
		DisplayName: "Rust Dev", Description: "Rust development tools (rust-analyzer, clippy)",
		Category: CategoryLanguage, IsDev: true, BaseLang: "rust",
		EnvFile:          "rust-dev.env",
		Requires:         []string{"rust", "cron"},
		VSCodeExtensions: []string{"rust-lang.rust-analyzer"},
	})

	r.add(&Feature{
		ID: "golang", BuildArg: "INCLUDE_GOLANG",
		DisplayName: "Go", Description: "Go toolchain (auto-resolves to latest patch)",
		Category: CategoryLanguage, VersionArg: "GO_VERSION", DefaultVersion: "1.23",
		EnvFile:      "golang.env",
		CacheVolumes: []string{"go-cache:/cache/go"},
	})
	r.add(&Feature{
		ID: "golang_dev", BuildArg: "INCLUDE_GOLANG_DEV",
		DisplayName: "Go Dev", Description: "Go development tools (gopls, dlv)",
		Category: CategoryLanguage, IsDev: true, BaseLang: "golang",
		EnvFile:          "golang.env",
		Requires:         []string{"golang"},
		VSCodeExtensions: []string{"golang.go"},
	})

	r.add(&Feature{
		ID: "ruby", BuildArg: "INCLUDE_RUBY",
		DisplayName: "Ruby", Description: "Ruby runtime (auto-resolves to latest patch)",
		Category: CategoryLanguage, VersionArg: "RUBY_VERSION", DefaultVersion: "3.4",
		EnvFile:      "ruby.env",
		CacheVolumes: []string{"bundle-cache:/cache/bundle"},
	})
	r.add(&Feature{
		ID: "ruby_dev", BuildArg: "INCLUDE_RUBY_DEV",
		DisplayName: "Ruby Dev", Description: "Ruby development tools (solargraph, rubocop)",
		Category: CategoryLanguage, IsDev: true, BaseLang: "ruby",
		EnvFile:          "ruby.env",
		Requires:         []string{"ruby"},
		VSCodeExtensions: []string{"shopify.ruby-lsp"},
	})

	r.add(&Feature{
		ID: "java", BuildArg: "INCLUDE_JAVA",
		DisplayName: "Java", Description: "Java JDK",
		Category: CategoryLanguage, VersionArg: "JAVA_VERSION", DefaultVersion: "21",
		EnvFile: "java.env",
	})
	r.add(&Feature{
		ID: "java_dev", BuildArg: "INCLUDE_JAVA_DEV",
		DisplayName: "Java Dev", Description: "Java development tools (jdtls)",
		Category: CategoryLanguage, IsDev: true, BaseLang: "java",
		EnvFile:          "java.env",
		Requires:         []string{"java"},
		VSCodeExtensions: []string{"redhat.java", "vscjava.vscode-java-debug"},
	})

	r.add(&Feature{
		ID: "r", BuildArg: "INCLUDE_R",
		DisplayName: "R", Description: "R statistical computing (auto-resolves to latest patch)",
		Category: CategoryLanguage, VersionArg: "R_VERSION", DefaultVersion: "4.4",
		EnvFile: "r.env",
	})
	r.add(&Feature{
		ID: "r_dev", BuildArg: "INCLUDE_R_DEV",
		DisplayName: "R Dev", Description: "R development tools (languageserver)",
		Category: CategoryLanguage, IsDev: true, BaseLang: "r",
		EnvFile:          "r.env",
		Requires:         []string{"r"},
		VSCodeExtensions: []string{"reditorsupport.r"},
	})

	r.add(&Feature{
		ID: "mojo", BuildArg: "INCLUDE_MOJO",
		DisplayName: "Mojo", Description: "Mojo programming language",
		Category: CategoryLanguage, VersionArg: "MOJO_VERSION", DefaultVersion: "25.4",
		EnvFile: "mojo.env",
	})
	r.add(&Feature{
		ID: "mojo_dev", BuildArg: "INCLUDE_MOJO_DEV",
		DisplayName: "Mojo Dev", Description: "Mojo development tools",
		Category: CategoryLanguage, IsDev: true, BaseLang: "mojo",
		EnvFile:  "mojo.env",
		Requires: []string{"mojo"},
	})

	r.add(&Feature{
		ID: "kotlin", BuildArg: "INCLUDE_KOTLIN",
		DisplayName: "Kotlin", Description: "Kotlin programming language (auto-installs Java, auto-resolves to latest patch)",
		Category: CategoryLanguage, VersionArg: "KOTLIN_VERSION", DefaultVersion: "2.3",
		EnvFile:      "kotlin.env",
		Requires:     []string{"java"},
		CacheVolumes: []string{"kotlin-cache:/cache/kotlin"},
	})
	r.add(&Feature{
		ID: "kotlin_dev", BuildArg: "INCLUDE_KOTLIN_DEV",
		DisplayName: "Kotlin Dev", Description: "Kotlin development tools (kotlin-language-server)",
		Category: CategoryLanguage, IsDev: true, BaseLang: "kotlin",
		EnvFile:          "kotlin.env",
		Requires:         []string{"kotlin", "java"},
		VSCodeExtensions: []string{"fwcd.kotlin"},
	})

	r.add(&Feature{
		ID: "android", BuildArg: "INCLUDE_ANDROID",
		DisplayName: "Android", Description: "Android SDK (auto-installs Java)",
		Category:     CategoryLanguage,
		EnvFile:      "android.env",
		Requires:     []string{"java"},
		CacheVolumes: []string{"android-cache:/cache/android-sdk", "gradle-cache:/cache/gradle"},
	})
	r.add(&Feature{
		ID: "android_dev", BuildArg: "INCLUDE_ANDROID_DEV",
		DisplayName: "Android Dev", Description: "Android emulator and system images",
		Category: CategoryLanguage, IsDev: true, BaseLang: "android",
		EnvFile:  "android.env",
		Requires: []string{"android", "java"},
	})

	// === Tools ===
	r.add(&Feature{
		ID: "dev_tools", BuildArg: "INCLUDE_DEV_TOOLS",
		DisplayName: "Dev Tools", Description: "General development tools (git extras, fzf, Claude Code CLI)",
		Category: CategoryTool,
		EnvFile:  "dev-tools.env",
		Requires: []string{"bindfs"},
		VSCodeExtensions: []string{
			"Anthropic.claude-code",
			"streetsidesoftware.code-spell-checker",
			"usernamehw.errorlens",
			"wayou.vscode-todo-highlight",
		},
	})
	r.add(&Feature{
		ID: "docker", BuildArg: "INCLUDE_DOCKER",
		DisplayName: "Docker", Description: "Docker CLI tools",
		Category:         CategoryTool,
		EnvFile:          "docker.env",
		VSCodeExtensions: []string{"ms-azuretools.vscode-docker"},
	})
	r.add(&Feature{
		ID: "op", BuildArg: "INCLUDE_OP",
		DisplayName: "1Password CLI", Description: "1Password CLI for secrets management",
		Category: CategoryTool,
		EnvFile:  "op-cli.env",
	})
	r.add(&Feature{
		ID: "cron", BuildArg: "INCLUDE_CRON",
		DisplayName: "Cron", Description: "Cron daemon (auto-enabled by rust_dev, dev_tools, bindfs)",
		Category:  CategoryTool,
		EnvFile:   "cron.env",
		ImpliedBy: []string{"rust_dev", "dev_tools", "bindfs"},
	})
	r.add(&Feature{
		ID: "bindfs", BuildArg: "INCLUDE_BINDFS",
		DisplayName: "Bindfs", Description: "FUSE permission overlay for macOS VirtioFS",
		Category:  CategoryTool,
		Requires:  []string{"cron"},
		ImpliedBy: []string{"dev_tools"},
	})
	r.add(&Feature{
		ID: "ollama", BuildArg: "INCLUDE_OLLAMA",
		DisplayName: "Ollama", Description: "Local LLM runtime",
		Category:     CategoryAI,
		EnvFile:      "ollama.env",
		CacheVolumes: []string{"ollama-cache:/cache/ollama"},
	})

	// === Cloud & Infrastructure ===
	r.add(&Feature{
		ID: "kubernetes", BuildArg: "INCLUDE_KUBERNETES",
		DisplayName: "Kubernetes", Description: "kubectl, k9s, Helm",
		Category: CategoryCloud,
		EnvFile:  "kubernetes.env",
	})
	r.add(&Feature{
		ID: "terraform", BuildArg: "INCLUDE_TERRAFORM",
		DisplayName: "Terraform", Description: "Terraform and related tools",
		Category:         CategoryCloud,
		EnvFile:          "terraform.env",
		VSCodeExtensions: []string{"hashicorp.terraform"},
	})
	r.add(&Feature{
		ID: "aws", BuildArg: "INCLUDE_AWS",
		DisplayName: "AWS CLI", Description: "AWS command-line interface",
		Category: CategoryCloud,
		EnvFile:  "aws.env",
	})
	r.add(&Feature{
		ID: "gcloud", BuildArg: "INCLUDE_GCLOUD",
		DisplayName: "Google Cloud SDK", Description: "Google Cloud command-line tools",
		Category: CategoryCloud,
		EnvFile:  "gcloud.env",
	})
	r.add(&Feature{
		ID: "cloudflare", BuildArg: "INCLUDE_CLOUDFLARE",
		DisplayName: "Cloudflare", Description: "Cloudflare Wrangler (requires Node.js)",
		Category: CategoryCloud,
		EnvFile:  "cloudflare.env",
		Requires: []string{"node"},
	})

	// === Database Clients ===
	r.add(&Feature{
		ID: "postgres_client", BuildArg: "INCLUDE_POSTGRES_CLIENT",
		DisplayName: "PostgreSQL Client", Description: "PostgreSQL client tools (psql)",
		Category: CategoryDatabase,
		EnvFile:  "database-clients.env",
	})
	r.add(&Feature{
		ID: "redis_client", BuildArg: "INCLUDE_REDIS_CLIENT",
		DisplayName: "Redis Client", Description: "Redis client tools (redis-cli)",
		Category: CategoryDatabase,
		EnvFile:  "database-clients.env",
	})
	r.add(&Feature{
		ID: "sqlite_client", BuildArg: "INCLUDE_SQLITE_CLIENT",
		DisplayName: "SQLite Client", Description: "SQLite client tools",
		Category: CategoryDatabase,
		EnvFile:  "database-clients.env",
	})

	return r
}

func (r *Registry) add(f *Feature) {
	r.features[f.ID] = f
	r.order = append(r.order, f.ID)
}

// Get returns a feature by ID, or nil if not found.
func (r *Registry) Get(id string) *Feature {
	return r.features[id]
}

// All returns all features in registration order.
func (r *Registry) All() []*Feature {
	result := make([]*Feature, 0, len(r.order))
	for _, id := range r.order {
		result = append(result, r.features[id])
	}
	return result
}

// ByCategory returns features matching the given category.
func (r *Registry) ByCategory(cat Category) []*Feature {
	var result []*Feature
	for _, id := range r.order {
		if r.features[id].Category == cat {
			result = append(result, r.features[id])
		}
	}
	return result
}

// Languages returns non-dev language features for the wizard language step.
func (r *Registry) Languages() []*Feature {
	var result []*Feature
	for _, f := range r.ByCategory(CategoryLanguage) {
		if !f.IsDev {
			result = append(result, f)
		}
	}
	return result
}

File: crates/containers-common/src/feature/registry.rs
//! Feature registry — the canonical ordered catalog of all container features.

use indexmap::IndexMap;

use super::{Category, Feature};

/// Holds the full catalog of available features in registration order.
#[derive(Debug, Clone)]
pub struct Registry {
    features: IndexMap<String, Feature>,
}

impl Registry {
    /// Creates the built-in feature registry populated from the containers
    /// build-args schema and hand-maintained metadata.
    #[must_use]
    #[expect(clippy::too_many_lines)]
    pub fn new() -> Self {
        let mut r = Self { features: IndexMap::new() };

        // === Languages ===

        r.add(Feature {
            id: "python".into(),
            build_arg: "INCLUDE_PYTHON".into(),
            display_name: "Python".into(),
            description: "Python runtime (auto-resolves to latest patch)".into(),
            category: Category::Language,
            version_arg: Some("PYTHON_VERSION".into()),
            default_version: Some("3.14".into()),
            env_file: Some("python.env".into()),
            cache_volumes: vec!["pip-cache:/cache/pip".into(), "poetry-cache:/cache/poetry".into()],
            ..Feature::default()
        });
        r.add(Feature {
            id: "python_dev".into(),
            build_arg: "INCLUDE_PYTHON_DEV".into(),
            display_name: "Python Dev".into(),
            description: "Python development tools (linters, formatters, LSP)".into(),
            category: Category::Language,
            is_dev: true,
            base_lang: Some("python".into()),
            env_file: Some("python-dev.env".into()),
            requires: vec!["python".into()],
            vscode_extensions: vec![
                "ms-python.python".into(),
                "ms-python.vscode-pylance".into(),
                "charliermarsh.ruff".into(),
            ],
            ..Feature::default()
        });

        r.add(Feature {
            id: "node".into(),
            build_arg: "INCLUDE_NODE".into(),
            display_name: "Node.js".into(),
            description: "Node.js runtime (auto-resolves to latest patch)".into(),
            category: Category::Language,
            version_arg: Some("NODE_VERSION".into()),
            default_version: Some("22".into()),
            env_file: Some("node.env".into()),
            cache_volumes: vec!["npm-cache:/cache/npm".into()],
            ..Feature::default()
        });
        r.add(Feature {
            id: "node_dev".into(),
            build_arg: "INCLUDE_NODE_DEV".into(),
            display_name: "Node.js Dev".into(),
            description: "Node.js development tools (LSP, debug)".into(),
            category: Category::Language,
            is_dev: true,
            base_lang: Some("node".into()),
            env_file: Some("node.env".into()),
            requires: vec!["node".into()],
            vscode_extensions: vec!["dbaeumer.vscode-eslint".into()],
            ..Feature::default()
        });

        r.add(Feature {
            id: "rust".into(),
            build_arg: "INCLUDE_RUST".into(),
            display_name: "Rust".into(),
            description: "Rust toolchain (auto-resolves to latest patch)".into(),
            category: Category::Language,
            version_arg: Some("RUST_VERSION".into()),
            default_version: Some("1.83".into()),
            env_file: Some("rust.env".into()),
            cache_volumes: vec![
                "cargo-cache:/cache/cargo".into(),
                "rustup-cache:/cache/rustup".into(),
            ],
            ..Feature::default()
        });
        r.add(Feature {
            id: "rust_dev".into(),
            build_arg: "INCLUDE_RUST_DEV".into(),
            display_name: "Rust Dev".into(),
            description: "Rust development tools (rust-analyzer, clippy)".into(),
            category: Category::Language,
            is_dev: true,
            base_lang: Some("rust".into()),
            env_file: Some("rust-dev.env".into()),
            requires: vec!["rust".into(), "cron".into()],
            vscode_extensions: vec!["rust-lang.rust-analyzer".into()],
            ..Feature::default()
        });

        r.add(Feature {
            id: "golang".into(),
            build_arg: "INCLUDE_GOLANG".into(),
            display_name: "Go".into(),
            description: "Go toolchain (auto-resolves to latest patch)".into(),
            category: Category::Language,
            version_arg: Some("GO_VERSION".into()),
            default_version: Some("1.23".into()),
            env_file: Some("golang.env".into()),
            cache_volumes: vec!["go-cache:/cache/go".into()],
            ..Feature::default()
        });
        r.add(Feature {
            id: "golang_dev".into(),
            build_arg: "INCLUDE_GOLANG_DEV".into(),
            display_name: "Go Dev".into(),
            description: "Go development tools (gopls, dlv)".into(),
            category: Category::Language,
            is_dev: true,
            base_lang: Some("golang".into()),
            env_file: Some("golang.env".into()),
            requires: vec!["golang".into()],
            vscode_extensions: vec!["golang.go".into()],
            ..Feature::default()
        });

        r.add(Feature {
            id: "ruby".into(),
            build_arg: "INCLUDE_RUBY".into(),
            display_name: "Ruby".into(),
            description: "Ruby runtime (auto-resolves to latest patch)".into(),
            category: Category::Language,
            version_arg: Some("RUBY_VERSION".into()),
            default_version: Some("3.4".into()),
            env_file: Some("ruby.env".into()),
            cache_volumes: vec!["bundle-cache:/cache/bundle".into()],
            ..Feature::default()
        });
        r.add(Feature {
            id: "ruby_dev".into(),
            build_arg: "INCLUDE_RUBY_DEV".into(),
            display_name: "Ruby Dev".into(),
            description: "Ruby development tools (solargraph, rubocop)".into(),
            category: Category::Language,
            is_dev: true,
            base_lang: Some("ruby".into()),
            env_file: Some("ruby.env".into()),
            requires: vec!["ruby".into()],
            vscode_extensions: vec!["shopify.ruby-lsp".into()],
            ..Feature::default()
        });

        r.add(Feature {
            id: "java".into(),
            build_arg: "INCLUDE_JAVA".into(),
            display_name: "Java".into(),
            description: "Java JDK".into(),
            category: Category::Language,
            version_arg: Some("JAVA_VERSION".into()),
            default_version: Some("21".into()),
            env_file: Some("java.env".into()),
            ..Feature::default()
        });
        r.add(Feature {
            id: "java_dev".into(),
            build_arg: "INCLUDE_JAVA_DEV".into(),
            display_name: "Java Dev".into(),
            description: "Java development tools (jdtls)".into(),
            category: Category::Language,
            is_dev: true,
            base_lang: Some("java".into()),
            env_file: Some("java.env".into()),
            requires: vec!["java".into()],
            vscode_extensions: vec!["redhat.java".into(), "vscjava.vscode-java-debug".into()],
            ..Feature::default()
        });

        r.add(Feature {
            id: "r".into(),
            build_arg: "INCLUDE_R".into(),
            display_name: "R".into(),
            description: "R statistical computing (auto-resolves to latest patch)".into(),
            category: Category::Language,
            version_arg: Some("R_VERSION".into()),
            default_version: Some("4.4".into()),
            env_file: Some("r.env".into()),
            ..Feature::default()
        });
        r.add(Feature {
            id: "r_dev".into(),
            build_arg: "INCLUDE_R_DEV".into(),
            display_name: "R Dev".into(),
            description: "R development tools (languageserver)".into(),
            category: Category::Language,
            is_dev: true,
            base_lang: Some("r".into()),
            env_file: Some("r.env".into()),
            requires: vec!["r".into()],
            vscode_extensions: vec!["reditorsupport.r".into()],
            ..Feature::default()
        });

        r.add(Feature {
            id: "mojo".into(),
            build_arg: "INCLUDE_MOJO".into(),
            display_name: "Mojo".into(),
            description: "Mojo programming language".into(),
            category: Category::Language,
            version_arg: Some("MOJO_VERSION".into()),
            default_version: Some("25.4".into()),
            env_file: Some("mojo.env".into()),
            ..Feature::default()
        });
        r.add(Feature {
            id: "mojo_dev".into(),
            build_arg: "INCLUDE_MOJO_DEV".into(),
            display_name: "Mojo Dev".into(),
            description: "Mojo development tools".into(),
            category: Category::Language,
            is_dev: true,
            base_lang: Some("mojo".into()),
            env_file: Some("mojo.env".into()),
            requires: vec!["mojo".into()],
            ..Feature::default()
        });

        r.add(Feature {
            id: "kotlin".into(),
            build_arg: "INCLUDE_KOTLIN".into(),
            display_name: "Kotlin".into(),
            description:
                "Kotlin programming language (auto-installs Java, auto-resolves to latest patch)"
                    .into(),
            category: Category::Language,
            version_arg: Some("KOTLIN_VERSION".into()),
            default_version: Some("2.3".into()),
            env_file: Some("kotlin.env".into()),
            requires: vec!["java".into()],
            cache_volumes: vec!["kotlin-cache:/cache/kotlin".into()],
            ..Feature::default()
        });
        r.add(Feature {
            id: "kotlin_dev".into(),
            build_arg: "INCLUDE_KOTLIN_DEV".into(),
            display_name: "Kotlin Dev".into(),
            description: "Kotlin development tools (kotlin-language-server)".into(),
            category: Category::Language,
            is_dev: true,
            base_lang: Some("kotlin".into()),
            env_file: Some("kotlin.env".into()),
            requires: vec!["kotlin".into(), "java".into()],
            vscode_extensions: vec!["fwcd.kotlin".into()],
            ..Feature::default()
        });

        r.add(Feature {
            id: "android".into(),
            build_arg: "INCLUDE_ANDROID".into(),
            display_name: "Android".into(),
            description: "Android SDK (auto-installs Java)".into(),
            category: Category::Language,
            env_file: Some("android.env".into()),
            requires: vec!["java".into()],
            cache_volumes: vec![
                "android-cache:/cache/android-sdk".into(),
                "gradle-cache:/cache/gradle".into(),
            ],
            ..Feature::default()
        });
        r.add(Feature {
            id: "android_dev".into(),
            build_arg: "INCLUDE_ANDROID_DEV".into(),
            display_name: "Android Dev".into(),
            description: "Android emulator and system images".into(),
            category: Category::Language,
            is_dev: true,
            base_lang: Some("android".into()),
            env_file: Some("android.env".into()),
            requires: vec!["android".into(), "java".into()],
            ..Feature::default()
        });

        // === Tools ===

        r.add(Feature {
            id: "dev_tools".into(),
            build_arg: "INCLUDE_DEV_TOOLS".into(),
            display_name: "Dev Tools".into(),
            description: "General development tools (git extras, fzf, Claude Code CLI)".into(),
            category: Category::Tool,
            env_file: Some("dev-tools.env".into()),
            requires: vec!["bindfs".into()],
            vscode_extensions: vec![
                "Anthropic.claude-code".into(),
                "streetsidesoftware.code-spell-checker".into(),
                "usernamehw.errorlens".into(),
                "wayou.vscode-todo-highlight".into(),
            ],
            ..Feature::default()
        });
        r.add(Feature {
            id: "docker".into(),
            build_arg: "INCLUDE_DOCKER".into(),
            display_name: "Docker".into(),
            description: "Docker CLI tools".into(),
            category: Category::Tool,
            env_file: Some("docker.env".into()),
            vscode_extensions: vec!["ms-azuretools.vscode-docker".into()],
            ..Feature::default()
        });
        r.add(Feature {
            id: "op".into(),
            build_arg: "INCLUDE_OP".into(),
            display_name: "1Password CLI".into(),
            description: "1Password CLI for secrets management".into(),
            category: Category::Tool,
            env_file: Some("op-cli.env".into()),
            ..Feature::default()
        });
        r.add(Feature {
            id: "cron".into(),
            build_arg: "INCLUDE_CRON".into(),
            display_name: "Cron".into(),
            description: "Cron daemon (auto-enabled by rust_dev, dev_tools, bindfs)".into(),
            category: Category::Tool,
            env_file: Some("cron.env".into()),
            implied_by: vec!["rust_dev".into(), "dev_tools".into(), "bindfs".into()],
            ..Feature::default()
        });
        r.add(Feature {
            id: "bindfs".into(),
            build_arg: "INCLUDE_BINDFS".into(),
            display_name: "Bindfs".into(),
            description: "FUSE permission overlay for macOS VirtioFS".into(),
            category: Category::Tool,
            requires: vec!["cron".into()],
            implied_by: vec!["dev_tools".into()],
            ..Feature::default()
        });
        r.add(Feature {
            id: "ollama".into(),
            build_arg: "INCLUDE_OLLAMA".into(),
            display_name: "Ollama".into(),
            description: "Local LLM runtime".into(),
            category: Category::Ai,
            env_file: Some("ollama.env".into()),
            cache_volumes: vec!["ollama-cache:/cache/ollama".into()],
            ..Feature::default()
        });

        // === Cloud & Infrastructure ===

        r.add(Feature {
            id: "kubernetes".into(),
            build_arg: "INCLUDE_KUBERNETES".into(),
            display_name: "Kubernetes".into(),
            description: "kubectl, k9s, Helm".into(),
            category: Category::Cloud,
            env_file: Some("kubernetes.env".into()),
            ..Feature::default()
        });
        r.add(Feature {
            id: "terraform".into(),
            build_arg: "INCLUDE_TERRAFORM".into(),
            display_name: "Terraform".into(),
            description: "Terraform and related tools".into(),
            category: Category::Cloud,
            env_file: Some("terraform.env".into()),
            vscode_extensions: vec!["hashicorp.terraform".into()],
            ..Feature::default()
        });
        r.add(Feature {
            id: "aws".into(),
            build_arg: "INCLUDE_AWS".into(),
            display_name: "AWS CLI".into(),
            description: "AWS command-line interface".into(),
            category: Category::Cloud,
            env_file: Some("aws.env".into()),
            ..Feature::default()
        });
        r.add(Feature {
            id: "gcloud".into(),
            build_arg: "INCLUDE_GCLOUD".into(),
            display_name: "Google Cloud SDK".into(),
            description: "Google Cloud command-line tools".into(),
            category: Category::Cloud,
            env_file: Some("gcloud.env".into()),
            ..Feature::default()
        });
        r.add(Feature {
            id: "cloudflare".into(),
            build_arg: "INCLUDE_CLOUDFLARE".into(),
            display_name: "Cloudflare".into(),
            description: "Cloudflare Wrangler (requires Node.js)".into(),
            category: Category::Cloud,
            env_file: Some("cloudflare.env".into()),
            requires: vec!["node".into()],
            ..Feature::default()
        });

        // === Database Clients ===

        r.add(Feature {
            id: "postgres_client".into(),
            build_arg: "INCLUDE_POSTGRES_CLIENT".into(),
            display_name: "PostgreSQL Client".into(),
            description: "PostgreSQL client tools (psql)".into(),
            category: Category::Database,
            env_file: Some("database-clients.env".into()),
            ..Feature::default()
        });
        r.add(Feature {
            id: "redis_client".into(),
            build_arg: "INCLUDE_REDIS_CLIENT".into(),
            display_name: "Redis Client".into(),
            description: "Redis client tools (redis-cli)".into(),
            category: Category::Database,
            env_file: Some("database-clients.env".into()),
            ..Feature::default()
        });
        r.add(Feature {
            id: "sqlite_client".into(),
            build_arg: "INCLUDE_SQLITE_CLIENT".into(),
            display_name: "SQLite Client".into(),
            description: "SQLite client tools".into(),
            category: Category::Database,
            env_file: Some("database-clients.env".into()),
            ..Feature::default()
        });

        r
    }

    fn add(&mut self, feature: Feature) {
        self.features.insert(feature.id.clone(), feature);
    }

    /// Returns a feature by ID, or `None` if not found.
    #[must_use]
    pub fn get(&self, id: &str) -> Option<&Feature> {
        self.features.get(id)
    }

    /// Returns all features in registration order.
    pub fn all(&self) -> impl Iterator<Item = &Feature> {
        self.features.values()
    }

    /// Returns features matching the given category, in registration order.
    pub fn by_category(&self, cat: Category) -> impl Iterator<Item = &Feature> {
        self.features.values().filter(move |f| f.category == cat)
    }

    /// Returns non-dev language features for the wizard language step.
    pub fn languages(&self) -> impl Iterator<Item = &Feature> {
        self.by_category(Category::Language).filter(|f| !f.is_dev)
    }
}

impl Default for Registry {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn registry_has_34_features() {
        let reg = Registry::new();
        assert_eq!(reg.features.len(), 34);
    }

    #[test]
    fn all_features_have_build_arg() {
        let reg = Registry::new();
        for f in reg.all() {
            assert!(!f.build_arg.is_empty(), "feature {} has empty build_arg", f.id);
        }
    }

    #[test]
    fn languages_excludes_dev() {
        let reg = Registry::new();
        for f in reg.languages() {
            assert!(!f.is_dev, "{} is a dev feature but appeared in languages()", f.id);
        }
    }

    #[test]
    fn dev_features_have_base_lang() {
        let reg = Registry::new();
        for f in reg.all() {
            if f.is_dev {
                assert!(f.base_lang.is_some(), "dev feature {} is missing base_lang", f.id);
            }
        }
    }

    #[test]
    fn requires_reference_valid_features() {
        let reg = Registry::new();
        for f in reg.all() {
            for req in &f.requires {
                assert!(
                    reg.get(req).is_some(),
                    "feature {} requires unknown feature {}",
                    f.id,
                    req
                );
            }
        }
    }

    #[test]
    fn implied_by_references_valid_features() {
        let reg = Registry::new();
        for f in reg.all() {
            for imp in &f.implied_by {
                assert!(reg.get(imp).is_some(), "feature {} has unknown implied_by {}", f.id, imp);
            }
        }
    }
}
